backpack/custom_module/graph_utils.py

```diff
@@ -0,0 +1,203 @@
+"""Transformation tools to make graph BackPACK compatible."""
+from copy import deepcopy
+from typing import Set
+
+from torch.fx import Graph, GraphModule, Node, Tracer
+from torch.nn import Flatten, Module
+
+from backpack.custom_module.branching import ActiveIdentity, Branch, SumModule
+from backpack.custom_module.scale_module import ScaleModule
+from backpack.utils import TORCH_VERSION_AT_LEAST_1_9_0
+
+
+class BackpackTracer(Tracer):
+    """Tracer that recognizes BackPACK's custom modules as 'leaf modules'."""
+
+    def is_leaf_module(
+        self, m: Module, module_qualified_name: str
+    ) -> bool:  # noqa: D102
+        if isinstance(m, (ScaleModule, SumModule, Branch, ActiveIdentity)):
+            return True
+        else:
+            return super().is_leaf_module(m, module_qualified_name)
+
+
+def print_table(module: Module) -> None:
+    """Prints a table of the module.
+
+    Args:
+        module: module to analyze
+    """
+    graph_resnet18 = BackpackTracer().trace(module)
+    graph_resnet18.print_tabular()
+
+
+def convert_module_to_backpack(module: Module) -> GraphModule:
+    """Convert all modules to BackPACK-compatible modules.
+
+    Transformations:
+    - mul -> ScaleModule
+    - add -> AddModule
+    - flatten -> nn.Flatten
+
+    Args:
+        module: module to convert
+
+    Returns:
+        backpack compatible module
+
+    Raises:
+        NotImplementedError: if not torch >= 1.9.0
+    """
+    if TORCH_VERSION_AT_LEAST_1_9_0 is False:
+        raise NotImplementedError(
+            "Conversion is only possible for torch >= 1.9.0. This is because these "
+            "functions use functionality such as torch.nn.Module.get_submodule"
+        )
+    print("\nMake module BackPACK compatible...")
+    module_new = _transform_mul_to_scale_module(module)
+    module_new = _transform_flatten_to_module(module_new)
+    module_new = _transform_add_to_sum_module(module_new)
+    module_new = _transform_inplace_to_normal(module_new)
+    module_new = _transform_remove_duplicates(module_new)
+    print("\tDelete unused modules.")
+    module_new.delete_all_unused_submodules()
+    print("Finished transformation.\n")
+    return module_new
+
+
+def _transform_mul_to_scale_module(module: Module) -> GraphModule:
+    print("\tBegin transformation: <built-in function mul> -> ScaleModule")
+    counter: int = 0
+    graph: Graph = BackpackTracer().trace(module)
+    for node in graph.nodes:
+        if node.op == "call_function":
+            if str(node.target) == "<built-in function mul>":
+                _change_node_to_module(
+                    node,
+                    "scale_module",
+                    module,
+                    ScaleModule(node.args[1]),
+                    (node.args[0],),
+                )
+            counter += 1
```

Comment:
I think the indentation is incorrect. Will refactor myself, please check correctness afterwards.

Reply:
Is correct now.